### PR TITLE
Change all APIs that returns `Result<String, Utf8Error>` to either ret `MagickCString` or `&'static CStr` to archive zero-copy and enable users to handle non utf-8 string

### PR DESCRIPTION
**Pull request by NobodyXu**
Plus:
 - [Improve doc for initialize](https://github.com/jmjoy/graphicsmagick-rs/commit/aeb9ba5c59cd07d98e5fd81f4cdb98aec7708020)
 - [Optimize str_to_c_string: from_vec_unchecked appends \0](https://github.com/jmjoy/graphicsmagick-rs/commit/cf79d22c9c80c2449c816db21cead68a98b29918)
 - [FIx: Call MagickFree on unwinding & refactor c_str_to_string](https://github.com/jmjoy/graphicsmagick-rs/commit/6e5ec82f027f47aebc207b78ca86d5350ff86d2b)
 - [FIx: Call MagickFree in c_arr_to_vec on unwinding](https://github.com/jmjoy/graphicsmagick-rs/commit/0022f234b6d83909d1bbbba5b69080f455265d27)
 - [Mark c_str_to_string* as unsafe](https://github.com/jmjoy/graphicsmagick-rs/commit/62fa57c305bb69b1fc9e9475e7085fa5d93e0be6)

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

### CODE REVIEW

File: graphicsmagick/src/wand/drawing.rs [hunk 1]
```diff
@@ -152,9 +152,8 @@ impl DrawingWand {
     ///
     /// must be deallocated by the user when it is no longer needed.
     ///
-    pub fn get_clip_path(&self) -> Result<String, Utf8Error> {
-        let c = unsafe { MagickDrawGetClipPath(self.wand) };
-        c_str_to_string(c)
+    pub fn get_clip_path(&self) -> Option<MagickCString> {
+        unsafe { MagickCString::new(MagickDrawGetClipPath(self.wand)) }
```
Comment (NobodyXu) [inline graphicsmagick/src/wand/drawing.rs:156]: @jmjoy Is it possible for these APIs that returns a c-style string to return `NULL`?

File: graphicsmagick/src/wand/magick.rs [hunk 2]
```diff
@@ -888,37 +886,33 @@ impl<'a> MagickWand<'a> {
...
 
     /// <http://www.graphicsmagick.org/wand/magick_wand.html#magickgetcopyright>
     ///
     /// MagickGetCopyright() returns the ImageMagick API copyright as a string.
     ///
-    pub fn get_copyright() -> Result<String, Utf8Error> {
-        let c = unsafe { MagickGetCopyright() };
-        c_str_to_string_no_free(c)
+    pub fn get_copyright() -> &'static CStr {
+        unsafe { CStr::from_ptr(MagickGetCopyright()) }
```
Comment (NobodyXu) [inline graphicsmagick/src/wand/magick.rs:899]: @jmjoy It is impossible for these functions, which probably returns a global const variable, to return `NULL` right?

File: graphicsmagick/src/wand/magick.rs [hunk 3]
```diff
@@ -2338,11 +2327,13 @@ impl<'a> MagickWand<'a> {
     ///
     /// MagickQueryFonts() returns any font that match the specified pattern.
     ///
-    pub fn query_fonts(pattern: &str) -> Option<Vec<Result<String, Utf8Error>>> {
+    pub fn query_fonts(pattern: &str) -> Option<Vec<Option<MagickCString>>> {
```
Comment (NobodyXu) [inline graphicsmagick/src/wand/magick.rs:2330]: Comment:

I will post another PR addressing functions returning a `Vec` to also make them zero-cost.

File: graphicsmagick/src/wand/drawing.rs (hunk 1 above)
Comment (NobodyXu) [reply graphicsmagick/src/wand/drawing.rs:156]: From the CI failure, it seems that this is absolutely possible.

File: graphicsmagick/src/wand/magick.rs (hunk 2 above)
Comment (NobodyXu) [reply graphicsmagick/src/wand/magick.rs]: I added debug assertion when constructing `&'static CStr` and it passed the CI tests, so maybe it is OK.

**Comment by NobodyXu** [comment]
@jmjoy This PR is ready for review.

File: graphicsmagick/src/utils.rs [hunk 4]
```diff
@@ -92,13 +99,60 @@ where
...
+#[derive(Debug)]
+pub struct MagickCString(Option<MagickAlloc>);
+
+impl MagickCString {
+    pub(crate) unsafe fn new(c: *const c_char) -> Self {
+        Self(MagickAlloc::new(c as *mut c_void))
+    }
+
+    /// Convert [`MagickCString`] to [`CStr`].
+    pub fn as_cstr(&self) -> &CStr {
```
Comment (jmjoy) [inline graphicsmagick/src/utils.rs:124]: Rename to `as_c_str`?

Comment (NobodyXu) [reply graphicsmagick/src/utils.rs]: Done!

File: graphicsmagick/src/wand/magick.rs (hunk 2 above)
Comment (jmjoy) [reply graphicsmagick/src/wand/magick.rs]: :ok_hand:

**Review by jmjoy — APPROVED** [review]
LGTM

**Pull request merged**